### PR TITLE
Save scroll position when scrolling.

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scroll_controller.dart [hunk 1]
```diff
@@ -180,7 +180,26 @@ class ScrollController extends ChangeNotifier {
...
+  ///
+  /// The arguments are generally passed to the [ScrollPosition] being created:
+  ///
+  ///  * `physics`: An instance of [ScrollPhysics] that determines how the
+  ///    [ScrollPosition] should react to user interactions, how it should
+  ///    simulate scrolling when released or flung, etc. The value will not be
+  ///    null. It typically comes from the [ScrollView] or other widget that
+  ///    creates the [Scrollable], or, if none was provided, from the ambient
+  ///    [ScrollConfiguration].
+  ///  * `context`: A [BuildContext] used for sending scroll notifications and
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_controller.dart]: Wrong type?

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 2]
```diff
@@ -195,15 +197,54 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
...
+  /// disposed then recreated in the same session. This might happen, for
+  /// instance, if a [ListView] is on one of the pages inside a [TabBarView],
+  /// and that page is displayed, then hidden, then displayed again.
+  ///
+  /// The default implementation reads the value from the nearest [PageStorage]
+  /// found from the [context]'s [ScrollContext.storageContext] property, and
+  /// sets it using [correctPixels], if [pixels] is still null.
+  @protected
+  void restoreScrollOffset() {
+    if (pixels == null) {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: Please test that the restored scroll offset wins over the initial scroll offset.

File: packages/flutter/lib/src/widgets/scroll_controller.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_controller.dart]: fixed

File: packages/flutter/lib/src/widgets/scroll_position.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: I adjusted the test to have a scroll controller set to 110.0.

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/8101

**Comment by Hixie** [comment]
PTAL.

**Comment by abarth** [comment]
LGTM

**Pull request merged**